src/transformers/training_args.py

```diff
@@ -968,6 +969,30 @@ def _no_sync_in_gradient_accumulation(self):
         """
         return not (self.deepspeed or is_sagemaker_dp_enabled() or is_sagemaker_mp_enabled())
 
+    @contextlib.contextmanager
+    def main_process_first(self):
```

Comment:
We should add a `local` bool argument here, to have a context manager that works:
- by letting local main processes first (default)
- by letting only the main process first

Reply:
Also maybe an optional description for the debug log statements?

Reply:
This?
```
local=True: each machine process 0
local=False: machine 0 process 0
```

may be `multi_node` is a more self-documenting name - i.e. that it's relevant to multi node setup?
```
multi_node=False: each machine process 0
multi_node=True: machine 0 process 0
```

Not attached... just not sure `local` is very clear as I had to ask you what you meant.

Reply:
> maybe an optional description for the debug log statements?

I'm not sure what you're suggesting here.

Reply:
`local` is what is used by PyTorch (`local_rank`) or PyTorch XLA, that's why I suggested it. For the `description` argument I was suggesting, it's an optional string you would add in your `logging.debug` (which are otherwise note super informative) that describes what is happening during that "lock". But it's not necessary.

Reply:
`local` it is then.

> For the description argument I was suggesting, it's an optional string you would add in your logging.debug (which are otherwise note super informative) that describes what is happening during that "lock". But it's not necessary.

This makes total sense! will add.

Reply:
done, please have a look. I also switched to using `self.local_process_index` and `self.process_index` to catch all the other frameworks - please correct me if I shouldn't have done this and stuck to pure pytorch api.